keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,442 @@
+- Resource overcommitment, fractional resource requirements, native compute resource (i.e., cpu and memory) with special metadata requirements, and group compute resources. They are out of our current scope.
+
+## Components
+### ResourceClass API
+During the initial phase, we propose to start with the following ResourceClass API spec that defines the basic ResourceClass name to the underlying node level compute resource plus metadata matching. We can extend this API to better support cluster resources and group resources in the following phases of development. However, this document will mostly focus on the design to support initial phase requirements.
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec ResourceClassSpec
+        // +optional
+}
+
+type ResourceClassSpec struct {
+        // raw resource name. E.g.: nvidia.com/gpu
+        ResourceName string
+        // ResourceSelector  selects resources. ORed from each selector.
+        resourceSelector []ResourcePropertySelector
+}
+
+type ResourcePropertySelector struct {
+        // A list of resource/device selector requirements
+        MatchExpressions []ResourceSelectorRequirement
+}
+
+// A resource selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values
+type ResourceSelectorRequirement struct {
+        // The label key that the selector applies to
+        Key string
+        // Example 0.1, intel etc
+        Values []string
+        // Similar to NodeSelectorOperator. Valid operators are: "In", "NotIn",
+        // "Exists", "DoesNotExist", "Gt", and "Lt".
+        Operator ResourceSelectorOperator
+}
+```
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  resourceName: "nvidia.com/gpu"
+  resourceSelector:
+    - matchExpressions:
+        - key: "memory"
+          operator: "Gt"
+          values:
+            - "15G"
+
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  containers:
+    - name: example-container
+      resources:
+        limits:
+          nvidia.high.mem: 2
+
+```
+Above resource class will select all the nvidia-gpus which have memory greater than 15GB.
+
+YAML example 2:
+```yaml
+kind: ResourceClass
+metadata:
+  name: fast.nic
+spec:
+  resourceName: "sfc.com/smartNIC"
+  labelSelector:
+    - matchExpressions:
+        - key: "speed"
+          operator: "Gt"
+          values:
+            - "40GBPS"
```

Comment:
Question to test my own comprehension here.  Could one reasonably have:

- key: "networkservice"
  operator: "Eq"
  values:
       - "radio-network"

Reply:
yes, key can be any attribute name advertised by device plugin.

Reply:
@vikaschoudhary16 OK, so what entity has to understand that "radio-network" is of type string, instead of type network bandwidth, or other type?